libnavigation-core/src/androidTest/java/com/mapbox/navigation/core/trip/service/ArtificialDriverTest.kt

```diff
@@ -0,0 +1,143 @@
+                val setRoutesResult =
+                    nativeNavigator.setRoutes(testRoute, reason = SetRoutesReason.NEW_ROUTE)
+                assertTrue("result is $setRoutesResult", setRoutesResult.isValue)
+                val statusesTracking = async<List<NavigationStatus>> {
+                    nativeNavigator.collectStatuses(untilRouteState = RouteState.COMPLETE)
+                }
+
+                for (location in events.map { it.location.mapToLocation() }) {
+                    assertTrue(nativeNavigator.updateLocation(location.toFixLocation()))
+                }
+
+                val states = statusesTracking.await()
+                val historyFile = suspendCoroutine<String> { continuation ->
+                    mapboxNavigation.historyRecorder.stopRecording {
+                        continuation.resume(it ?: "null")
+                    }
+                }
+                val notTrackingStates = states.filter { it.routeState != RouteState.TRACKING }
+                assertTrue(
+                    "not all states are tracking, " +
+                        "see history file $historyFile: $notTrackingStates",
+                    notTrackingStates.isEmpty()
+                )
+            }
+        }
+}
+
+private fun createArtificialLocationUpdates(
+    testRoute: NavigationRoute
+): List<ReplayEventUpdateLocation> {
+    val replayRouteMapper = ReplayRouteMapper()
+    return replayRouteMapper
+        .mapDirectionsRouteGeometry(testRoute.directionsRoute)
+        .filterIsInstance<ReplayEventUpdateLocation>()
+}
+
+private suspend fun MapboxNativeNavigator.collectStatuses(
+    untilRouteState: RouteState
+): MutableList<NavigationStatus> {
+    val statues = mutableListOf<NavigationStatus>()
+    statusUpdates()
+        .map { it.status }
+        .takeWhile { it.routeState != untilRouteState }
+        .toList(statues)
+    return statues
+}
+
+data class OnStatusUpdateParameters(
+    val origin: NavigationStatusOrigin,
+    val status: NavigationStatus
+)
+
+@OptIn(ExperimentalCoroutinesApi::class)
+fun MapboxNativeNavigator.statusUpdates(): Flow<OnStatusUpdateParameters> {
+    return callbackFlow {
+        val observer = NavigatorObserver { origin, status ->
+            this.trySend(OnStatusUpdateParameters(origin, status))
+        }
+        addNavigatorObserver(observer)
+        awaitClose {
+            removeNavigatorObserver(observer)
+        }
+    }
+}
+
+private suspend fun withNavigators(
+    block: suspend (MapboxNavigation, MapboxNativeNavigator) -> Unit
+) {
+    val context = InstrumentationRegistry.getInstrumentation().targetContext
+    val mapboxNavigation = MapboxNavigationProvider.create(
+        NavigationOptions.Builder(context)
+            .accessToken(context.getString(R.string.mapbox_access_token))
+            .build()
+    )
+    block(mapboxNavigation, MapboxNativeNavigatorImpl)
+    mapboxNavigation.onDestroy()
```

Comment:
`onDestroyed` won't be called if the test fails, you should use a rule for that.

Reply:
thanks. I think try-finaly will work too, won't they? 250aa56b1a8a328b72e5e80a5a34176148fcc08b

Reply:
They will, yes.